lqns/lqns.1
.\" t -*- nroff -*-
.TH lqns 1 "23 February 2023" "5.28.2"
.\" $Id: lqns.1 16441 2023-02-23 21:39:47Z greg $
.\"
.\" --------------------------------
.SH "NAME"
lqns \- solve layered queueing network models.
.SH "SYNOPSIS"
.br
.B lqns
[\fIOPTIONS\fR].\|.\|. [\fIFILE\fR] \&.\|.\|.
.SH "DESCRIPTION"
\fBLqns\fP is used to solve layered queueing network models using 
analytic techniques.  Models can be specified using the LQN modeling
language, or with extensible markup language (XML).  Refer to the
\fI``Layered Queueing Network Solver and Simulator User Manual''\fP
for details of the model and for a complete description of the input file
formats for the program.
.PP
\fBLqns\fP reads its input from \fIfilename\fR, specified at the
command line if present, or from the standard input otherwise.  By
default, output for an input file \fIfilename\fR specified on the
command line will be placed in the file \fIfilename\fB.out\fR.  If the
\fB\-p\fP switch is used, parseable output will also be written into
\fIfilename\fB.p\fR. If XML input or the \fB\-x\fP switch is used, XML output will be written to 
\fIfilename\fB.lqxo\fR.  This behaviour can be changed using the
\fB\-o\fP\fIoutput\fR switch, described below.  If several files are
named, then each is treated as a separate model and output will be
placed in separate output files.  If input is from the standard input,
output will be directed to the standard output.  The file name `\fI-\fR' is
used to specify standard input.
.PP
The \fB\-o\fP\fIoutput\fR option can be used to direct output to the file
\fIoutput\fR regardless of the source of input.  Output will be XML
if XML input or if the \fB\-x\fP switch is used, parseable output if the \fB\-p\fP switch is used,
and normal output otherwise.  Multiple input files cannot be specified
when using this option.  Output can be directed to standard output by
using \fB\-o\fP\fI-\fR (i.e., the output file name is `\fI-\fR'.)
.SH "OPTIONS"
.TP
\fB\-a\fP, \fB\-\-no-advisories\fR
Ignore advisories.  The default is to print out all advisories.
.TP
\fB\-b\fP, \fB\-\-bounds-only\fR
This option is used to compute the ``Type 1 throughput bounds'' only.
These bounds are computed assuming no contention anywhere in the model
and represent the guaranteed not to exceed values.
.TP
\fB\-c\fP, \fB\-\-convergence-value\fR=\fIarg\fR
Set the convergence value to \fIarg\fP.  
\fIArg\fP must be a number between 0.0 and 1.0.
.TP
\fB\-d\fP, \fB\-\-debug\fR=\fIarg\fR
This option is used to enable debug output.
\fIArg\fP can be one of:
.RS
.TP
\fBall\fR
Enable all debug output.
.TP
\fBforks\fR
Print out the fork-join matching process.
.TP
\fBinterlock\fR
Print out the interlocking table and the interlocking between all tasks and processors.
.TP
\fBlqx\fR
Print out the actions the LQX parser while reading an LQX program.
.TP
\fBovertaking\fR
Print the overtaking probabilities in the output file.
.TP
\fBsubmodels\fR
Print out the contents of all of the submodels found in the model.
.TP
\fBvariance\fR
Print out variance calculation.
.TP
\fBxml\fR
Print out the actions of the Expat parser while reading XML input.
.RE
.TP
\fB\-e\fP, \fB\-\-error\fR=\fIarg\fR
This option is to enable floating point exception handling.
\fIArg\fP must be one of the following:
.RS
.TP 3
\fBa\fP
Abort immediately on a floating point error (provided the floating point unit can do so).
.TP 3
\fBd\fP
Abort on floating point errors. (default)
.TP 3
\fBi\fP
Ignore floating point errors.
.TP 3
\fBw\fP
Warn on floating point errors.
The solver checks for floating point overflow, division by zero and invalid operations.
Underflow and inexact result exceptions are always ignored.
.PP
In some instances, infinities  will be propogated within the solver.  Please refer to the
\fBstop-on-message-loss\fP pragma below.
.RE
.TP
\fB\-f\fP, \fB\-\-fast-linearizer\fR
This option is used to set options for quick solution of a model using One-Step (Bard-Schweitzer) MVA.
It is equivalent to setting \fBpragma\fP \fImva\fP=\fIone-step\fP, \fIlayering\fP=\fIbatched\fP, \fImultiserver\fP=\fIconway\fP
.TP
\fB\-h\fP, \fB\-\-huge\fR
Solve using one-step-schweitzer, no interlocking, and Rolia multiserver.
.TP
\fB\-H\fP, \fB\-\-help\fR=\fIarg\fR
Print a summary of the command line options.  The optional argument shows help for -d, -t, -z, and -P respectively.
.TP
\fB\-i\fP, \fB\-\-iteration-limit\fR=\fIarg\fR
Set the maximum number of iterations to \fIarg\fP.
\fIArg\fP must be an integer greater than 0.  The default value is 50.
.TP
\fB\-I\fP, \fB\-\-input-format\fR=\fIarg\fR
This option is used to force the input file format to either \fIxml\fP or \fIlqn\fP.
By default, if the suffix of the input filename is one of: \fI.in\fP, \fI.lqn\fP or \fI.xlqn\fP
, then the LQN parser will be used.  Otherwise, input is assumed to be XML.
.TP
\fB\-j\fP, \fB\-\-json\fR
Generate JSON output regardless of input format.
.TP
\fB\-M\fP, \fB\-\-mol-underrelaxation\fR=\fIarg\fR
Set the under-relaxation factor to \fIarg\fP for the MOL (Rolia) multiserver approximation.  If the approximation is failing, lower this value.
\fIArg\fP must be a number between 0.0 and 1.0.
The default value is 0.5.
.TP
\fB\-n\fP, \fB\-\-no-execute\fR
Build the model, but do not solve.
The input is checked for validity but no output is generated.
.TP
\fB\-o\fP, \fB\-\-output\fR=\fIarg\fR
Direct analysis results to \fIoutput\fP.  A filename of `\fI-\fR'
directs output to standard output.  If \fIoutput\fR is a directory, all output is saved in
\fIoutput/input.out\fR. If the input model contains a SPEX program with loops, the SPEX output is sent to
\fIoutput\fR; the individual model output files are found in the directory
\fIoutput.d\fR. If \fBlqns\fP is invoked with this
option, only one input file can be specified.
.TP
\fB\-p\fP, \fB\-\-parseable\fR
Generate parseable output suitable as input to other programs such as
\fBlqn2ps(1)\fP and \fBsrvndiff(1)\fP.  If input is from
\fIfilename\fR, parseable output is directed to \fIfilename\fB.p\fR.
If standard input is used for input, then the parseable output is sent
to the standard output device.  If the \fB\-o\fP\fIoutput\fR option is used, the
parseable output is sent to the file name \fIoutput\fR.
(In this case, only parseable output is emitted.)
.TP
\fB\-P\fP, \fB\-\-pragma\fR=\fIarg\fR
Change the default solution strategy.  Refer to the PRAGMAS section
below for more information.
.TP
\fB\-r\fP, \fB\-\-rtf\fR
Output results using Rich Text Format instead of plain text.  Processors, entries and tasks with high utilizations are coloured in red.
.TP
\fB\-t\fP, \fB\-\-trace\fR=\fIarg\fR
This option is used to set tracing  options which are used to print out various
intermediate results  while a model is being solved.
\fIarg\fP can be any combination of the following:
.RS
.TP
\fBactivities\fR
Print out results of activity aggregation.
.TP
\fBconvergence\fR=\fIarg\fR
Print out convergence value after each submodel is solved.
This option is useful for tracking the rate of convergence for a model.
The optional numeric argument supplied to this option will print out the convergence value for the specified MVA submodel, otherwise,
the convergence value for all submodels will be printed.
.TP
\fBdelta-wait\fR
Print out difference in entry service time after each submodel is solved.
.TP
\fBforks\fR
Print out overlap table for forks prior to submodel solution.
.TP
\fBidle-time\fR
Print out computed idle time after each submodel is solved.
.TP
\fBinterlock\fR
Print out interlocking adjustment before each submodel is solved.
.TP
\fBintermediate\fR
Print out intermediate solutions at the print interval specified in the model.
The print interval field in the input is ignored otherwise.
.TP
\fBjoins\fR
Print out computed join delay and join overlap table prior to submodel solution.
.TP
\fBmva\fR=\fIarg\fR
Output the inputs and results of each MVA submodel for every iteration of the solver.
The optional argument is a bit set of the submodels to output.  Submodel 1 is 0x1, 
submodel 2 is 0x2, submodel 3 is 0x4, etc.  By default all submodels are traced.
.TP
\fBovertaking\fR
Print out overtaking calculations.
.TP
\fBquorum\fR
Print quorum traces.
.TP
\fBreplication\fR

.TP
\fBthroughput\fR
Print throughput's values.
.TP
\fBvariance\fR
Print out the variances calculated after each submodel is solved.
.TP
\fBvirtual-entry\fR
Print waiting time for each rendezvous in the model after it has been computed; include virtual entries.
.TP
\fBwait\fR
Print waiting time for each rendezvous in the model after it has been computed.
.RE
.TP
\fB\-u\fP, \fB\-\-underrelaxation\fR=\fIarg\fR
Set the underrelaxation to \fIarg\fP.
\fIArg\fP must be a number between 0.0 and 1.0.
The default value is 0.9.
.TP
\fB\-v\fP, \fB\-\-verbose\fR
Generate output after each iteration of the MVA solver and the convergence value at the end of each outer iteration of the solver.
.TP
\fB\-V\fP, \fB\-\-version\fR
Print out version and copyright information.
.TP
\fB\-w\fP, \fB\-\-no-warnings\fR
Ignore warnings.  The default is to print out all warnings.
.TP
\fB\-x\fP, \fB\-\-xml\fR
Generate XML output regardless of input format.
.TP
\fB\-z\fP, \fB\-\-special\fR=\fIarg\fR
This option is used to select special options.  Arguments of the form
\fInn\fP are integers while arguments of the form \fInn.n\fP are real
numbers.  \fIArg\fP can be any of the following:
.RS
.TP
\fBfull-reinitialize\fR
For multiple runs, reinitialize all service times at processors.
.TP
\fBgenerate\fR=\fIarg\fR

NEWS
* Bugs fixes in 5.28
297 	lqiolib 	Gnuplot and result enhancements
343 	qnsolver 	Improve compatibility with Java Modelling Tools 
356 	miscella 	rpm build issues.
359 	lqiolib 	Expressions are not output by Spex::print_header().
366 	lqn2xxx 	EXC_BAD_ACCESS Layer::transmorgrify() at layer.cc:571:17 [opt] 
411 	lqiolib 	JMVA issues.
412 	qnsolver 	Save JMVA WhatIf results.
413 	Document 	Missing plainurl. 	
415 	petrisrv 	Signal SIGSEGV in LQIO::DOM::Phase::getServiceTime () at dom_phase.cpp:88 
416 	lqiolib 	Fixes for plotting population mix WhatIf output. 
417 	qnsolver 	Bounds solution issues. 
418 	lqsim 		lqsim crashes running -t... 	
419 	qnsolver 	model.cc:(.text+0x18ff): undefined reference to `typeinfo for QNIO::JMVA_Document'

* Bugs fixed in 5.27
192 	libmva 	 	MVA::marginalProbabilities2 mis-dimensioned. 
356 	miscella 	rpm build issues.
373 	lqns 	 	Processor waiting times at Rolia multiserver make no sense. 
380 	lqns 	 	lqns hangs with Conway multiserver. 
382 	lqn2xxx 	lqn2ps -Oout doesn't print processors. 
383 	lqiolib 	<pre> precedences can accept more than one activity in XML (JSON?) input. 
385 	petrisrv 	V5 petrisrvn: faulty models causing aborts.
386 	lqn2csv 	getopt_long headaches. 
387 	srvndiff 	File glob not working. 
389 	srvndiff 	C:\msys64\mingw32\bin\srvndiff.exe: model-001.lqxo:12: error: unclosed token 
390 	lqn2xxx 	lqn2xxx: command not found. 
391 	lqsim 		lqsim/regression/Makefile.am, 5*-replication. 
392 	petrisrv 	Calls are not being constructed
393 	lqiolib 	Save (and print) marginal queue probabilities.
395 	petrisrv 	V5 petrisrvn: faulty models causing aborts or failing.
397 	lqiolib 	Add qnap2 parser for BCMP models.
398 	lqiolib 	Refactor LQX math handing from ExternalVariables to LQX Expressions.
399 	lqiolib 	WhatIf construction is bogus.
400 	lqn2xxx 	Converting from json to lqn fails on Windows.
401 	libmva 		libc++abi: terminating with uncaught exception of type std::invalid_argument: default-output
402 	miscella 	Mac M1 issues.
403 	libmva 	 	43-activities.lqnx: Floating point exception: mva.cc 541: Inexact result, Invalid operation
405 	lqns 	 	Optimizer bug: option.h
406 	libmva 	 	Refactor fpgoop.
407 	lqns 	 	--srvn-layering crashes in V6 interlocking code
409 	petrisrv 	Petrisrvn fails when SPEX/LQX is used to alter the arrival rate to an entry. 
411 	lqiolib 	JMVA issues. 

* Bugs fixed in 5.26
365 	lqns	 	The warning for zero service time should be suppressed for reference tasks with think time. 
369 	petrisrv 	Segmentation violation with --overtaking 
370 	lqiolib 	Activity parameters probabilities and rates are constants. 
379 	lqns	 	V5 Regression. 	
372 	lqiolib 	Result variables can be reused. 
374 	lqns 		Asynchronous Cycles #1 	
375 	lqns 		Asynchronous Cycles #1 	
376 	lqn2xxx 	Overloaded infinite servers with open arrivals... 
378 	lqn2xxx 	Activity throughput not initialized. 

* Bugs fixed in 5.25
338 	lqns 		Queuing time not equal to zero when NCust = NServers for Zhou multiserver. 
352 	lqn2xxx 	--no-xxxx doesn't work. 
353 	lqns 		Assertion failed: (ix && ix <= sz), function operator[], file vector.h, line 174. 
357 	para_lib 	parasol/include/sem*caphore.h conflicts with /usr/include/semaphore.h 
360 	lqngen	 	lqngen -L<n> doesn't work. 
361 	lqn2xxx 	EXC_BAD_ACCESS (code=EXC_I386_GPFLT) LQIO::DOM::Document::~Document(this=0x0000000100a04080) at dom_document.cpp:123:3 
362 	lqn2csv 	terminate called after throwing an instance of 'std::runtime_error' 
363 	lqn2csv 	vector::_M_range_check: __n (which is 18446744073709551615) >= this->size() (which is 3) 

* Bugs fixed in 5.24
321 	lqiolib 	Runtime Exception Occured: Object of type phase does not have property service_time 
328 	spex 		Using $1, $2... as result variables fails. 
339 	lqiolib 	--reload-lqx segvs or generally is borked. 
340 	lqiolib 	loadResults not implemented in Json_Document. 
341 	lqn2xxx 	lqn2svg -Q1 -o 01-sanity-queue.svg -Zprune 01-sanity.in aborted. 
342 	qnsolver 	Assertion failed: (L[n][0] == nullptr), function reset, file mva.cc, line 353. 
344 	lqiolib 	Automatic variable creation generates duplicate symbols for 2-class models. 
345 	libmva 		No results for NCust = [X,0] 
346 	libmva 		class(station("m"),"k").throughput fails. 
347 	lqns 		Changes needed to -zgenerate 
351 	lqn2xxx 	EMF output borked. 

* Bugs fixes in 5.23
300 	lqiolib 	Station results not computed for single class models. 	
324 	lqns 		Runtime Exception Occured: Batch_Model::addToSubmodel: bad submodel for task t1 
329 	qnsolver 	uncaught exception of type std::out_of_range: map::at: key not found 
331 	lqn2xxx 	Rendering issues with activities. 
332 	lqsim 		Lqsim CFS: 71-fair, 72-fair 
271 	lqn2xxx 	Group by regex may or may not work. 

lqns/lqns.cc
/*  -*- c++ -*-
 * $Id: lqns.cc 16440 2023-02-23 21:03:40Z greg $
 *
 * Command line processing.
 *
 * Copyright the Real-Time and Distributed Systems Group,
 * Department of Systems and Computer Engineering,
 * Carleton University, Ottawa, Ontario, Canada. K1S 5B6
 *
 * November, 1994
 *
 * ------------------------------------------------------------------------
 */

#include "lqns.h"
#include <sstream>
#include <fstream>
#include <cstdio>
#include <cstdlib>
#include <cmath>
#include <errno.h>
#include <fcntl.h>           /* Definition of AT_* constants */
#if HAVE_GETOPT_H
#include <getopt.h>
#endif
#if HAVE_LIBGEN_H
#include <libgen.h>
#endif
#include <sys/stat.h>
#include <unistd.h>
#include <lqio/commandline.h>
#include <lqio/filename.h>
#include <lqio/input.h>
#include <lqio/srvn_spex.h>
#include <lqio/dom_pragma.h>
#if !HAVE_GETSUBOPT
#include <lqio/getsbopt.h>
#endif
#include <mva/fpgoop.h>
#include "errmsg.h"
#include "generate.h"
#include "help.h"
#include "flags.h"
#include "model.h"
#include "option.h"
#include "pragma.h"

extern "C" int srvndebug;

static char copyrightDate[20];

/* -- */

struct FLAGS flags;

#if HAVE_GETOPT_LONG
const struct option longopts[] =
    /* name */ /* has arg */ /*flag */ /* val */
{
    { "no-advisories",				no_argument,	   nullptr, 'a' },
    { "bounds-only",				no_argument,	   nullptr, 'b' },
    { LQIO::DOM::Pragma::_convergence_value_,	required_argument, nullptr, 'c' },
    { "debug",					required_argument, nullptr, 'd' },
    { "error",					required_argument, nullptr, 'e' },
    { LQIO::DOM::Pragma::_fast_,		no_argument,	   nullptr, 'f' },
    { "huge",					no_argument,	   nullptr, 'h' },
    { "help",					optional_argument, nullptr, 'H' },
    { LQIO::DOM::Pragma::_iteration_limit_,	required_argument, nullptr, 'i' },
    { "input-format",				required_argument, nullptr, 'I' },
    { "json",					no_argument,	   nullptr, 'j' },
    { LQIO::DOM::Pragma::_mol_underrelaxation_, required_argument, nullptr, 'M' },
    { "no-execute",				no_argument,	   nullptr, 'n' },
    { "output",					required_argument, nullptr, 'o' },
    { "parseable",				no_argument,	   nullptr, 'p' },
    { "pragma",					required_argument, nullptr, 'P' },
    { "rtf",					no_argument,	   nullptr, 'r' },
    { "trace",					required_argument, nullptr, 't' },
    { LQIO::DOM::Pragma::_underrelaxation_,	required_argument, nullptr, 'u' },
    { "verbose",				no_argument,	   nullptr, 'v' },
    { "version",				no_argument,	   nullptr, 'V' },
    { "no-warnings",				no_argument,	   nullptr, 'w' },
    { "xml",					no_argument,	   nullptr, 'x' },
    { "special",				required_argument, nullptr, 'z' },
    { LQIO::DOM::Pragma::_exact_,		no_argument,	   nullptr, 256+'e' },
    { LQIO::DOM::Pragma::_schweitzer_,		no_argument,	   nullptr, 256+'s' },
    { "batch-layering",				no_argument,	   nullptr, 256+'b' },	/* NOP */
    { "hwsw-layering",				no_argument,	   nullptr, 256+'h' },
    { "method-of-layers",			no_argument,	   nullptr, 256+'m' },
    { "squashed-layering",			no_argument,	   nullptr, 256+'z' },
    { "srvn-layering",				no_argument,	   nullptr, 256+'l' },
    { "processor-sharing",			no_argument,	   nullptr, 256+'p' },
#if HAVE_LIBGSL && HAVE_LIBGSLCBLAS
    { "quorum",					no_argument,	   nullptr, 256+'q' },
#endif
    { "no-stop-on-message-loss",		no_argument,	   nullptr, 256+'o' },	/* Ignore open */
    { "no-variance",				no_argument,	   nullptr, 256+'v' },
    { "reload-lqx",				no_argument,	   nullptr, 512+'r' },
    { "restart",				no_argument,	   nullptr, 512+'R' },
    { "no-header",				no_argument,	   nullptr, 512+'h' },
    { LQIO::DOM::Pragma::_spex_convergence_,	required_argument, nullptr, 256+'C' },
    { "print-comment",				no_argument,	   nullptr, 512+'c' },
    { "print-interval",				optional_argument, nullptr, 512+'p' },
    { "reset-mva",				no_argument,	   nullptr, 256+'r' },
    { "trace-mva",				optional_argument, nullptr, 256+'t' },
    { "debug-submodels",			no_argument,	   nullptr, 256+'S' },
    { "debug-json",				no_argument,	   nullptr, 512+'j' },
    { "debug-lqx",				no_argument,	   nullptr, 512+'l' },
    { "debug-srvn",				no_argument,	   nullptr, 512+'y' },
    { "debug-xml",				no_argument,	   nullptr, 512+'x' },
    { "print-lqx",				no_argument,	   nullptr, 512+'s' },
    { nullptr, 0, nullptr, 0 }
};
#else
const struct option * longopts = nullptr;
#endif
const char opts[]       = "abc:d:e:fhH:i:I:jM:no:pP:rt:u:vVwxz:";

static void init_flags ();

#if (defined(linux) || defined(__linux__)) && !defined(__USE_XOPEN_EXTENDED)
extern "C" int getsubopt (char **, char * const *, char **);
#endif

/*
 * Main line.
 */

int main (int argc, char *argv[])
{
    std::string outputFileName = "";
    LQIO::DOM::Document::OutputFormat output_format = LQIO::DOM::Document::OutputFormat::DEFAULT;
    LQIO::CommandLine command_line( longopts );
    Options::Debug::initialize();
    Options::Trace::initialize();
    Options::Special::initialize();
    Model::solve_using solve_function = &Model::compute;

    unsigned global_error_flag = 0;     /* Error detected anywhere??    */

    char * options;

    LQIO::io_vars.init( VERSION, basename( argv[0] ), LQIO::severity_action );
    std::copy( local_error_messages.begin(), local_error_messages.end(), std::inserter( LQIO::error_messages, LQIO::error_messages.begin() ) );
    
    command_line = LQIO::io_vars.lq_toolname;

    sscanf( "$Date: 2023-02-23 16:03:40 -0500 (Thu, 23 Feb 2023) $", "%*s %s %*s", copyrightDate );

    matherr_disposition = fp_exception_reporting::DEFERRED_ABORT;

    init_flags();

    pragmas.insert( getenv( "LQNS_PRAGMAS" ) );

    LQIO::DOM::DocumentObject::setSeverity(LQIO::ERR_NO_QUANTUM_SCHEDULING, LQIO::error_severity::WARNING );	// Don't care for lqns.

    for ( ;; ) {
#if HAVE_GETOPT_LONG
        const int c = getopt_long( argc, argv, opts, longopts, nullptr );
#else
        const int c = getopt( argc, argv, opts );
#endif
        if ( c == EOF) break;

        command_line.append( c, optarg );

	try {
	    switch ( c ) {
	    case 'a':
		pragmas.insert( LQIO::DOM::Pragma::_severity_level_, LQIO::DOM::Pragma::_warning_ );
		break;

	    case 'b':
		flags.bounds_only = true;
		break;

	    case 256+'b':
		break;

	    case 'c':
		pragmas.insert( LQIO::DOM::Pragma::_convergence_value_, optarg != nullptr ? optarg : std::string("") );
		break;

	    case 256+'C':
		pragmas.insert( LQIO::DOM::Pragma::_spex_convergence_, optarg != nullptr ? optarg : std::string("") );
		break;
		
	    case 512+'c':
		/* Set immediately, as it can't be changed once the SPEX program is loaded */
		LQIO::Spex::__print_comment = true;
		pragmas.insert(LQIO::DOM::Pragma::_spex_comment_,"true");
		break;

	    case 'd':
		options = optarg;
		while ( *options ) {
		    char * value = nullptr;
		    const char * subopt = options;
		    const int ix = getsubopt( &options, Options::Debug::__options.data(), &value );
		    if ( ix >= 0 ) {
			Options::Debug::exec( ix, (value == nullptr ? std::string("") : value) );
		    } else {
			throw std::invalid_argument( std::string("--debug=") + subopt );
		    }
		}
		break;

	    case 'e':                       /* Error handling.      */
		switch ( optarg[0] ) {
		case 'a':
		    matherr_disposition = fp_exception_reporting::IMMEDIATE_ABORT;
		    break;

		case 'd':
		    matherr_disposition = fp_exception_reporting::DEFERRED_ABORT;
		    break;

		case 'i':
		    matherr_disposition = fp_exception_reporting::IGNORE;
		    break;

		case 'w':
		    matherr_disposition = fp_exception_reporting::REPORT;
		    break;

		default:
		    std::cerr << LQIO::io_vars.lq_toolname << ": invalid argument to -e -- " << optarg << std::endl;
		    break;
		}
		break;

	    case 256+'e':
		pragmas.insert(LQIO::DOM::Pragma::_mva_,LQIO::DOM::Pragma::_exact_);
		break;

	    case 'f':
		pragmas.insert(LQIO::DOM::Pragma::_layering_,LQIO::DOM::Pragma::_batched_);
		pragmas.insert(LQIO::DOM::Pragma::_mva_,LQIO::DOM::Pragma::_one_step_linearizer_);
		pragmas.insert(LQIO::DOM::Pragma::_multiserver_,LQIO::DOM::Pragma::_conway_);
		break;

	    case 'h':
		pragmas.insert(LQIO::DOM::Pragma::_interlocking_,LQIO::DOM::Pragma::_no_);
		pragmas.insert(LQIO::DOM::Pragma::_mva_,LQIO::DOM::Pragma::_one_step_);
		pragmas.insert(LQIO::DOM::Pragma::_multiserver_,LQIO::DOM::Pragma::_rolia_);
		break;

	    case 'H':
		usage( optarg );
		exit(0);

	    case 256+'h':
		pragmas.insert(LQIO::DOM::Pragma::_layering_,LQIO::DOM::Pragma::_hwsw_);
		break;

	    case 512+'h':
		/* Set immediately, as it can't be changed once the SPEX program is loaded */
		LQIO::Spex::__no_header = true;
		pragmas.insert(LQIO::DOM::Pragma::_spex_header_,"false");
		break;

	    case 'I':
		if ( strcasecmp( optarg, "xml" ) == 0 ) {
		    Model::__input_format = LQIO::DOM::Document::InputFormat::XML;
		} else if ( strcasecmp( optarg, "lqn" ) == 0 ) {
		    Model::__input_format = LQIO::DOM::Document::InputFormat::LQN;
		} else if ( strcasecmp( optarg, "json" ) == 0 ) {
		    Model::__input_format = LQIO::DOM::Document::InputFormat::JSON;
		} else {
		    std::cerr << LQIO::io_vars.lq_toolname << ": invalid argument to -I -- " << optarg << std::endl;
		    exit( 1 );
		}
		break;

	    case 'i':
		pragmas.insert( LQIO::DOM::Pragma::_iteration_limit_, optarg != nullptr ? optarg : std::string("") );
		break;

	    case 'j':
		output_format = LQIO::DOM::Document::OutputFormat::JSON;
		break;

	    case 256+'l':
		pragmas.insert(LQIO::DOM::Pragma::_layering_,LQIO::DOM::Pragma::_srvn_);
		break;

	    case (512+'l'):
		LQIO::DOM::Document::lqx_parser_trace(stderr);
		break;

	    case 'M':
		pragmas.insert( LQIO::DOM::Pragma::_mol_underrelaxation_, optarg != nullptr ? optarg : std::string("") );
		break;
		
	    case 256+'m':
		pragmas.insert(LQIO::DOM::Pragma::_layering_,LQIO::DOM::Pragma::_mol_);
		break;

	    case 'n':
		flags.no_execute = true;
		break;

	    case 'o':
		outputFileName = optarg;
		break;

	    case 256+'o':
		pragmas.insert(LQIO::DOM::Pragma::_stop_on_message_loss_,LQIO::DOM::Pragma::_no_);
		break;

	    case 'p':
		output_format = LQIO::DOM::Document::OutputFormat::PARSEABLE;
		break;

	    case 'P':       /* Pragma processing... */
		if ( !pragmas.insert( optarg ) ) {
		    Pragma::usage( std::cerr );
		    exit( INVALID_ARGUMENT );
		}
		break;

	    case 256+'p':
		pragmas.insert( LQIO::DOM::Pragma::_processor_scheduling_, LQIO::SCHEDULE::PS );
		break;

	    case 512+'p':
		Options::Special::print_interval( optarg != nullptr ? optarg : std::string("") );
		break;

	    case 256+'q': //tomari quorum options
		flags.disable_expanding_quorum_tree = true;
		break;

	    case 'r':
		flags.rtf_output = true;
		break;

	    case 256+'r':
		flags.reset_mva = true;
		break;

	    case 512+'r':
		solve_function = &Model::reload;
		break;

	    case 512+'R':
		solve_function = &Model::restart;
		break;

	    case 256+'s':
		pragmas.insert( LQIO::DOM::Pragma::_mva_, LQIO::DOM::Pragma::_schweitzer_ );
		break;

	    case 256+'S':
		Options::Debug::submodels( optarg != nullptr ? optarg : std::string("") );
		break;
		    
	    case 512+'s':
		flags.print_lqx = true;
		break;

	    case 't':
		options = optarg;
		while ( *options ) {
		    char * value = nullptr;
		    const char * subopt = options;
		    const int ix = getsubopt( &options, Options::Trace::__options.data(), &value );
		    if ( ix >= 0 ) {
			Options::Trace::exec( ix, (value != nullptr ? value : std::string("") ) );
		    } else {
			throw std::invalid_argument( std::string("--trace=") + subopt );
		    }
		}
		break;

	    case 256+'t':
		Options::Trace::mva( optarg != nullptr ? optarg : std::string("") );
		break;

	    case 'u':
		pragmas.insert( LQIO::DOM::Pragma::_underrelaxation_, optarg != nullptr ? optarg : std::string("") );
		break;

	    case 'v':
		Options::Trace::verbose( optarg != nullptr ? optarg : std::string("") );
		break;

	    case 'V':
		std::cout << "Layered Queueing Network Analyser, Version " << VERSION << std::endl << std::endl;
		std::cout << "  Copyright " << copyrightDate << " the Real-Time and Distributed Systems Group," << std::endl;
		std::cout << "  Department of Systems and Computer Engineering," << std::endl;
		std::cout << "  Carleton University, Ottawa, Ontario, Canada. K1S 5B6" << std::endl << std::endl;
		break;

	    case (256+'v'):
		pragmas.insert(LQIO::DOM::Pragma::_variance_, LQIO::DOM::Pragma::_none_);
		break;

	    case 'w':
		pragmas.insert(LQIO::DOM::Pragma::_severity_level_,LQIO::DOM::Pragma::_run_time_);
		break;

	    case 'x':
		output_format = LQIO::DOM::Document::OutputFormat::XML;
		break;

	    case 512+'x':
		LQIO::DOM::Document::__debugXML = true;
		break;

	    case 512+'y':
		srvndebug = true;
		break;

	    case 'z':
		options = optarg;
		while ( *options ) {
		    char * value = nullptr;
		    const char * subopt = options;
		    const int ix = getsubopt( &options, Options::Special::__options.data(), &value );
		    if ( ix >= 0 ) {
			Options::Special::exec( ix, (value == nullptr ? "" : value) );
		    } else {
			throw std::invalid_argument( std::string("--special=") + subopt );
		    }
		}
		break;

	    case (256+'z'):
		pragmas.insert(LQIO::DOM::Pragma::_layering_,LQIO::DOM::Pragma::_squashed_);
		break;

	    default:
		usage();
		break;
	    }
	}
	catch ( const std::invalid_argument& e )
	{
	    std::cerr << LQIO::io_vars.lq_toolname << ": invalid argument to " << e.what() << "." << std::endl;
	    exit( 1 );
	}
    }
    LQIO::io_vars.lq_command_line = command_line.c_str();

    if ( flags.generate ) {
	if ( flags.no_execute ) {
	    std::cerr << LQIO::io_vars.lq_toolname << ": -n is incompatible with -zgenerate.  -zgenerate ignored." << std::endl;
	} else if ( access( Generate::__directory_name.c_str(), R_OK|W_OK|X_OK ) < 0 && ENOENT ) {
#if defined(__WINNT__)
	    int rc = mkdir( Generate::__directory_name.c_str() );
#else
	    int rc = mkdir( Generate::__directory_name.c_str(), S_IRUSR|S_IWUSR|S_IXUSR|S_IRGRP|S_IWGRP|S_IXGRP|S_IWOTH|S_IROTH|S_IXOTH );
#endif
	    if ( rc < 0 ) {
		std::cerr << LQIO::io_vars.lq_toolname << ": Cannot create directory " << Generate::__directory_name << ": " << strerror( errno ) << "." << std::endl;
	    }
	}
    }

    /* Process all command line arguments.  If none specified, then     */
    /* input is assumed to come in from stdin.                          */

    if ( optind == argc ) {

        global_error_flag = Model::solve( solve_function, "-", outputFileName, output_format );

    } else {

        const int file_count = argc - optind;           /* Number of files on cmd line  */

        if ( file_count > 1 ) {
            if ( LQIO::Filename::isFileName( outputFileName ) && LQIO::Filename::isDirectory( outputFileName ) == 0 ) {
                std::cerr << LQIO::io_vars.lq_toolname << ": Too many input files specified with the option: -o"
                     << outputFileName
                     << std::endl;
                exit( INVALID_ARGUMENT );
            }
            if ( !Generate::__directory_name.empty() ) {
                std::cerr << LQIO::io_vars.lq_toolname << ": Too many input files specified with the option: -zgenerate="
                     << Generate::__directory_name
                     << std::endl;
                exit( INVALID_ARGUMENT );
            }
        }

        for ( ; optind < argc; ++optind ) {
            if ( file_count > 1 ) {
                std::cout << argv[optind] << ':' << std::endl;
            }
            global_error_flag |= Model::solve( solve_function, argv[optind], outputFileName, output_format );
        }
    }

    return global_error_flag;
}



/*
 * Initialize flags to their default values;
 */

static
void init_flags()
{
    flags.no_execute            = false;
    flags.bounds_only           = false;
    flags.rtf_output            = false;
    flags.generate              = false;
    flags.reset_mva		= false;
    flags.print_overtaking      = false;
    flags.single_step           = false;
    flags.print_lqx		= false;

    flags.trace_activities      = false;
    flags.trace_convergence     = false;
    flags.trace_customers	= false;
    flags.trace_forks           = false;
    flags.trace_idle_time       = false;
    flags.trace_interlock       = false;
    flags.trace_intermediate    = false;
    flags.trace_joins           = false;
    flags.trace_overtaking      = false;
    flags.trace_replication     = false;
    flags.trace_virtual_entry   = false;
    flags.trace_wait            = false;

#if HAVE_LIBGSL && HAVE_LIBGSLCBLAS
    flags.min_steps             = 7;            /* Default of 2 steps. */
#else
    flags.min_steps             = 2;            /* Default of 2 steps. */
#endif

    flags.ignore_overhanging_threads = false;
    flags.full_reinitialize          = false;               /* Maybe a pragma?                      */
}

/*
 * Common underrelaxation code.
 */

double
under_relax( const double old_value, const double new_value, const double relax )
{
    if ( std::isfinite( new_value ) && std::isfinite( old_value ) ) {
	return new_value * relax + old_value * (1.0 - relax);
    } else {
	return new_value;
    }
}
